File: flutter/pulsar/ios/Classes/PulsarPlugin.swift
import Flutter
import UIKit

public class PulsarPlugin: NSObject, FlutterPlugin {
  private let pulsar = Pulsar()
  private lazy var realtimeComposer = pulsar.getRealtimeComposer()
  private var patternComposers: [Int: PatternComposer] = [:]
  private var nextPatternComposerId = 1

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "pulsar", binaryMessenger: registrar.messenger())
    let instance = PulsarPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    let args = call.arguments as? [String: Any]
    switch call.method {

    // MARK: — Pulsar
    case "Pulsar_play":
      guard let name = args?["name"] as? String else {
        result(FlutterError(code: "INVALID_ARGS", message: "name required", details: nil))
        return
      }
      pulsar.getPresets().getByName(name)?.play()
      result(nil)

    case "Pulsar_enableHaptics":
      guard let state = args?["state"] as? Bool else {
        result(FlutterError(code: "INVALID_ARGS", message: "state required", details: nil))
        return
      }
      pulsar.enableHaptics(state: state)
      result(nil)

    case "Pulsar_enableSound":
      guard let state = args?["state"] as? Bool else {
        result(FlutterError(code: "INVALID_ARGS", message: "state required", details: nil))
        return
      }
      pulsar.enableSound(state: state)
      result(nil)

    case "Pulsar_enableCache":
      guard let state = args?["state"] as? Bool else {
        result(FlutterError(code: "INVALID_ARGS", message: "state required", details: nil))
        return
      }
      pulsar.enableCache(state: state)
      result(nil)

    case "Pulsar_isCacheEnabled":
      result(pulsar.getPresets().isCacheEnabled())

    case "Pulsar_clearCache":
      pulsar.clearCache()
      result(nil)

    case "Pulsar_preloadPreset":
      guard let name = args?["presetName"] as? String else {
        result(FlutterError(code: "INVALID_ARGS", message: "presetName required", details: nil))
        return
      }
      pulsar.getPresets().preloadPresetByName(name)
      result(nil)

    case "Pulsar_presetExists":
      guard let name = args?["name"] as? String else {
        result(FlutterError(code: "INVALID_ARGS", message: "name required", details: nil))
        return
      }
      result(pulsar.getPresets().getByName(name) != nil)

    case "Pulsar_preloadPresets":
      guard let names = args?["presetNames"] as? [String] else {
        result(FlutterError(code: "INVALID_ARGS", message: "presetNames required", details: nil))
        return
      }
      pulsar.preloadPresets(presetNames: names)
      result(nil)

    case "Pulsar_stopHaptics":
      pulsar.stopHaptics()
      result(nil)

    case "Pulsar_shutDownEngine":
      pulsar.shutDownEngine()
      result(nil)

    case "Pulsar_isHapticsEnabled":
      result(pulsar.isHapticsEnabled)

    case "Pulsar_canPlayHaptics":
      result(pulsar.canPlayHaptics())

    case "Pulsar_hapticSupport":
      // iOS currently models support as binary for Flutter: standard support or none.
      let supported = pulsar.isHapticsSupported()
      result(supported ? 2 : 0)

    case "Pulsar_forceHapticsSupportLevel":
      // No-op on iOS — CoreHaptics support level is hardware-determined.
      result(nil)

    case "Pulsar_enableImpulseCompositionMode":
      // No-op on iOS.
      result(nil)

    case "Pulsar_setRealtimeComposerStrategy":
      // No-op on iOS.
      result(nil)

    // MARK: — RealtimeComposer
    case "RealtimeComposer_set":
      guard let amplitude = args?["amplitude"] as? Double,
            let frequency = args?["frequency"] as? Double else {
        result(FlutterError(code: "INVALID_ARGS", message: "amplitude and frequency required", details: nil))
        return
      }
      if let strategyError = applyRealtimeStrategy(args) {
        result(strategyError)
        return
      }
      realtimeComposer.set(amplitude: Float(amplitude), frequency: Float(frequency))
      result(nil)

    case "RealtimeComposer_stop":
      if let strategyError = applyRealtimeStrategy(args) {
        result(strategyError)
        return
      }
      realtimeComposer.stop()
      result(nil)

    case "RealtimeComposer_isActive":
      if let strategyError = applyRealtimeStrategy(args) {
        result(strategyError)
        return
      }
      result(realtimeComposer.isActive)

    case "RealtimeComposer_playDiscrete":
      guard let amplitude = args?["amplitude"] as? Double,
            let frequency = args?["frequency"] as? Double else {
        result(FlutterError(code: "INVALID_ARGS", message: "amplitude and frequency required", details: nil))
        return
      }
      if let strategyError = applyRealtimeStrategy(args) {
        result(strategyError)
        return
      }
      realtimeComposer.playDiscrete(amplitude: Float(amplitude), frequency: Float(frequency))
      result(nil)

    // MARK: — PatternComposer
    case "PatternComposer_parsePattern":
      guard let data = args?["data"] as? [String: Any],
            let patternData = parsePatternData(data) else {
        result(FlutterError(code: "INVALID_ARGS", message: "valid pattern data required", details: nil))
        return
      }
      let composerId = args?["composerId"] as? Int
      let resolvedComposerId = composerId ?? nextPatternComposerId
      if composerId == nil {
        nextPatternComposerId += 1
      }
      let patternComposer = patternComposers[resolvedComposerId] ?? {
        let composer = pulsar.getPatternComposer()
        patternComposers[resolvedComposerId] = composer
        return composer
      }()
      patternComposer.parsePattern(hapticsData: patternData)
      result(resolvedComposerId)

    case "PatternComposer_playPattern":
      guard let data = args?["data"] as? [String: Any],
            let patternData = parsePatternData(data) else {
        result(FlutterError(code: "INVALID_ARGS", message: "valid pattern data required", details: nil))
        return
      }
      let composerId = args?["composerId"] as? Int
      let resolvedComposerId = composerId ?? nextPatternComposerId
      if composerId == nil {
        nextPatternComposerId += 1
      }
      let patternComposer = patternComposers[resolvedComposerId] ?? {
        let composer = pulsar.getPatternComposer()
        patternComposers[resolvedComposerId] = composer
        return composer
      }()
      patternComposer.playPattern(hapticsData: patternData)
      result(resolvedComposerId)

    case "PatternComposer_play":
      guard let composerId = args?["composerId"] as? Int else {
        result(FlutterError(code: "INVALID_ARGS", message: "composerId required", details: nil))
        return
      }
      patternComposers[composerId]?.play()
      result(nil)

    case "PatternComposer_playAudioOnly":
      guard let composerId = args?["composerId"] as? Int else {
        result(FlutterError(code: "INVALID_ARGS", message: "composerId required", details: nil))
        return
      }
      patternComposers[composerId]?.playAudioOnly()
      result(nil)

    case "PatternComposer_stop":
      guard let composerId = args?["composerId"] as? Int else {
        result(FlutterError(code: "INVALID_ARGS", message: "composerId required", details: nil))
        return
      }
      patternComposers[composerId]?.stop()
      result(nil)

    case "PatternComposer_release":
      guard let composerId = args?["composerId"] as? Int else {
        result(FlutterError(code: "INVALID_ARGS", message: "composerId required", details: nil))
        return
      }
      patternComposers.removeValue(forKey: composerId)?.stop()
      result(nil)

    default:
      result(FlutterMethodNotImplemented)
    }
  }

  // MARK: — Helpers

  private func parsePatternData(_ data: [String: Any]) -> PatternData? {
    guard let continuous = data["continuousPattern"] as? [String: Any],
          let amplitudeRaw = continuous["amplitude"] as? [[String: Any]],
          let frequencyRaw = continuous["frequency"] as? [[String: Any]],
          let discreteRaw = data["discretePattern"] as? [[String: Any]] else {
      return nil
    }

    let amplitude = amplitudeRaw.compactMap { pt -> ValuePoint? in
      guard let time = pt["time"] as? Double, let value = pt["value"] as? Double else { return nil }
      return ValuePoint(time: time, value: Float(value))
    }
    let frequency = frequencyRaw.compactMap { pt -> ValuePoint? in
      guard let time = pt["time"] as? Double, let value = pt["value"] as? Double else { return nil }
      return ValuePoint(time: time, value: Float(value))
    }
    let discrete = discreteRaw.compactMap { pt -> DiscretePoint? in
      guard let time = pt["time"] as? Double,
            let amp = pt["amplitude"] as? Double,
            let freq = pt["frequency"] as? Double else { return nil }
      return DiscretePoint(time: time, amplitude: Float(amp), frequency: Float(freq))
    }

    return PatternData(
      continuousPattern: ContinuousPattern(amplitude: amplitude, frequency: frequency),
      discretePattern: discrete
    )
  }

  private func applyRealtimeStrategy(_ args: [String: Any]?) -> FlutterError? {
    guard let strategy = args?["strategy"] else {
      return nil
    }
    guard strategy is Int else {
      return FlutterError(code: "INVALID_ARGS", message: "invalid strategy", details: nil)
    }
    return nil
  }
}

File: flutter/pulsar/android/src/main/kotlin/com/swmansion/pulsar/PulsarPlugin.kt
package com.swmansion.pulsar

import android.app.Activity
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import com.swmansion.pulsar.composers.PatternComposer
import com.swmansion.pulsar.types.CompatibilityMode
import com.swmansion.pulsar.types.ConfigPoint
import com.swmansion.pulsar.types.ContinuousPattern
import com.swmansion.pulsar.types.PatternData
import com.swmansion.pulsar.types.RealtimeComposerStrategy
import com.swmansion.pulsar.types.ValuePoint

/** PulsarPlugin */
class PulsarPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {

    private lateinit var channel: MethodChannel
    private var activity: Activity? = null
    private var pulsar: Pulsar? = null
    private val patternComposers = mutableMapOf<Int, PatternComposer>()
    private var nextPatternComposerId = 1

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "pulsar")
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        pulsar = Pulsar(binding.activity)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
        pulsar = null
        patternComposers.clear()
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
        pulsar = Pulsar(binding.activity)
    }

    override fun onDetachedFromActivity() {
        activity = null
        pulsar = null
        patternComposers.clear()
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        val p = pulsar
        if (p == null) {
            result.error("NOT_INITIALIZED", "Pulsar is not initialized (no Activity)", null)
            return
        }

        when (call.method) {
            "Pulsar_play" -> {
                val name = call.argument<String>("name")
                    ?: return result.error("INVALID_ARGS", "name required", null)
                p.getPresets().getByName(name)?.play()
                result.success(null)
            }

            "Pulsar_enableHaptics" -> {
                val state = call.argument<Boolean>("state")
                    ?: return result.error("INVALID_ARGS", "state required", null)
                p.enableHaptics(state)
                result.success(null)
            }

            "Pulsar_enableSound" -> {
                val state = call.argument<Boolean>("state")
                    ?: return result.error("INVALID_ARGS", "state required", null)
                p.enableSound(state)
                result.success(null)
            }

            "Pulsar_enableCache" -> {
                val state = call.argument<Boolean>("state")
                    ?: return result.error("INVALID_ARGS", "state required", null)
                p.enableCache(state)
                result.success(null)
            }

            "Pulsar_isCacheEnabled" -> {
                result.success(p.isCacheEnabled())
            }

            "Pulsar_clearCache" -> {
                p.clearCache()
                result.success(null)
            }

            "Pulsar_preloadPreset" -> {
                val name = call.argument<String>("presetName")
                    ?: return result.error("INVALID_ARGS", "presetName required", null)
                p.preloadPreset(name)
                result.success(null)
            }

            "Pulsar_presetExists" -> {
                val name = call.argument<String>("name")
                    ?: return result.error("INVALID_ARGS", "name required", null)
                result.success(p.getPresets().getByName(name) != null)
            }

            "Pulsar_preloadPresets" -> {
                val names = call.argument<List<String>>("presetNames")
                    ?: return result.error("INVALID_ARGS", "presetNames required", null)
                p.preloadPresets(names)
                result.success(null)
            }

            "Pulsar_stopHaptics" -> {
                p.stopHaptics()
                result.success(null)
            }

            "Pulsar_shutDownEngine" -> {
                result.success(null)
            }

            "Pulsar_isHapticsEnabled" -> {
                result.success(p.isHapticsEnabled())
            }

            "Pulsar_canPlayHaptics" -> {
                result.success(p.canPlayHaptics())
            }

            "Pulsar_hapticSupport" -> {
                result.success(p.hapticSupport().ordinal)
            }

            "Pulsar_forceHapticsSupportLevel" -> {
                val level = call.argument<Int>("level")
                    ?: return result.error("INVALID_ARGS", "level required", null)
                val mode = CompatibilityMode.entries.getOrNull(level)
                    ?: return result.error("INVALID_ARGS", "invalid level", null)
                p.forceHapticsSupportLevel(mode)
                result.success(null)
            }

            "Pulsar_enableImpulseCompositionMode" -> {
                val state = call.argument<Boolean>("state")
                    ?: return result.error("INVALID_ARGS", "state required", null)
                p.enableImpulseCompositionMode(state)
                result.success(null)
            }

            "Pulsar_setRealtimeComposerStrategy" -> {
                val strategy = call.argument<Int>("strategy")
                    ?: return result.error("INVALID_ARGS", "strategy required", null)
                val s = RealtimeComposerStrategy.entries.getOrNull(strategy)
                    ?: return result.error("INVALID_ARGS", "invalid strategy", null)
                p.realtimeComposerStrategy = s
                result.success(null)
            }

            "RealtimeComposer_set" -> {
                val amplitude = call.argument<Double>("amplitude")?.toFloat()
                    ?: return result.error("INVALID_ARGS", "amplitude required", null)
                val frequency = call.argument<Double>("frequency")?.toFloat()
                    ?: return result.error("INVALID_ARGS", "frequency required", null)
                val strategyIndex = call.argument<Int>("strategy")
                val strategy = if (strategyIndex != null) {
                    RealtimeComposerStrategy.entries.getOrNull(strategyIndex)
                        ?: return result.error("INVALID_ARGS", "invalid strategy", null)
                } else {
                    null
                }
                p.getRealtimeComposer(strategy).set(amplitude, frequency)
                result.success(null)
            }

            "RealtimeComposer_stop" -> {
                val strategyIndex = call.argument<Int>("strategy")
                val strategy = if (strategyIndex != null) {
                    RealtimeComposerStrategy.entries.getOrNull(strategyIndex)
                        ?: return result.error("INVALID_ARGS", "invalid strategy", null)
                } else {
                    null
                }
                p.getRealtimeComposer(strategy).stop()
                result.success(null)
            }

            "RealtimeComposer_isActive" -> {
                val strategyIndex = call.argument<Int>("strategy")
                val strategy = if (strategyIndex != null) {
                    RealtimeComposerStrategy.entries.getOrNull(strategyIndex)
                        ?: return result.error("INVALID_ARGS", "invalid strategy", null)
                } else {
                    null
                }
                result.success(p.getRealtimeComposer(strategy).isActive())
            }

            "RealtimeComposer_playDiscrete" -> {
                val amplitude = call.argument<Double>("amplitude")?.toFloat()
                    ?: return result.error("INVALID_ARGS", "amplitude required", null)
                val frequency = call.argument<Double>("frequency")?.toFloat()
                    ?: return result.error("INVALID_ARGS", "frequency required", null)
                val strategyIndex = call.argument<Int>("strategy")
                val strategy = if (strategyIndex != null) {
                    RealtimeComposerStrategy.entries.getOrNull(strategyIndex)
                        ?: return result.error("INVALID_ARGS", "invalid strategy", null)
                } else {
                    null
                }
                p.getRealtimeComposer(strategy).playDiscrete(amplitude, frequency)
                result.success(null)
            }

            "PatternComposer_parsePattern" -> {
                val data = call.argument<Map<String, Any>>("data")
                    ?: return result.error("INVALID_ARGS", "data required", null)
                val patternData = parsePatternData(data)
                    ?: return result.error("INVALID_ARGS", "invalid pattern data", null)
                val composerId = call.argument<Int>("composerId")
                val resolvedComposerId = composerId ?: nextPatternComposerId++
                val composer = patternComposers[resolvedComposerId] ?: p.getPatternComposer().also {
                    patternComposers[resolvedComposerId] = it
                }
                composer.parsePattern(patternData)
                result.success(resolvedComposerId)
            }

            "PatternComposer_playPattern" -> {
                val data = call.argument<Map<String, Any>>("data")
                    ?: return result.error("INVALID_ARGS", "data required", null)
                val patternData = parsePatternData(data)
                    ?: return result.error("INVALID_ARGS", "invalid pattern data", null)
                val composerId = call.argument<Int>("composerId")
                val resolvedComposerId = composerId ?: nextPatternComposerId++
                val composer = patternComposers[resolvedComposerId] ?: p.getPatternComposer().also {
                    patternComposers[resolvedComposerId] = it
                }
                composer.parsePattern(patternData)
                composer.play()
                result.success(resolvedComposerId)
            }

            "PatternComposer_play" -> {
                val composerId = call.argument<Int>("composerId")
                    ?: return result.error("INVALID_ARGS", "composerId required", null)
                patternComposers[composerId]?.play()
                result.success(null)
            }

            "PatternComposer_playAudioOnly" -> {
                val composerId = call.argument<Int>("composerId")
                    ?: return result.error("INVALID_ARGS", "composerId required", null)
                patternComposers[composerId]?.playAudioOnly()
                result.success(null)
            }

            "PatternComposer_stop" -> {
                val composerId = call.argument<Int>("composerId")
                    ?: return result.error("INVALID_ARGS", "composerId required", null)
                patternComposers[composerId]?.stop()
                result.success(null)
            }

            "PatternComposer_release" -> {
                val composerId = call.argument<Int>("composerId")
                    ?: return result.error("INVALID_ARGS", "composerId required", null)
                patternComposers.remove(composerId)?.stop()
                result.success(null)
            }

            else -> result.notImplemented()
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun parsePatternData(data: Map<String, Any>): PatternData? {
        val continuous = data["continuousPattern"] as? Map<String, Any> ?: return null
        val amplitudeRaw = continuous["amplitude"] as? List<Map<String, Any>> ?: return null
        val frequencyRaw = continuous["frequency"] as? List<Map<String, Any>> ?: return null
        val discreteRaw = data["discretePattern"] as? List<Map<String, Any>> ?: return null

        val amplitude = amplitudeRaw.mapNotNull { pt ->
            val time = (pt["time"] as? Number)?.toLong() ?: return@mapNotNull null
            val value = (pt["value"] as? Number)?.toFloat() ?: return@mapNotNull null
            ValuePoint(time, value)
        }
        val frequency = frequencyRaw.mapNotNull { pt ->
            val time = (pt["time"] as? Number)?.toLong() ?: return@mapNotNull null
            val value = (pt["value"] as? Number)?.toFloat() ?: return@mapNotNull null
            ValuePoint(time, value)
        }
        val discrete = discreteRaw.mapNotNull { pt ->
            val time = (pt["time"] as? Number)?.toLong() ?: return@mapNotNull null
            val amp = (pt["amplitude"] as? Number)?.toFloat() ?: return@mapNotNull null
            val freq = (pt["frequency"] as? Number)?.toFloat() ?: return@mapNotNull null
            ConfigPoint(time, amp, freq)
        }

        return PatternData(
            continuousPattern = ContinuousPattern(amplitude, frequency),
            discretePattern = discrete
        )
    }
}
